File: 10. Backtracking/078-subsets.py
# Leetcode 78

# Given an integer array nums of unique elements, 
# return all possible subsets (the power set).
# The solution set must not contain duplicate subsets. 
# Return the solution in any order.

from typing import List


def subsets(nums: List[int]) -> List[int]:
    result = []
    subset = []

    def dfs(i): 
        if i >= len(nums):
            result.append(subset.copy())
            return
        # decision to include
        subset.append(nums[i])
        dfs(i + 1)
        # decision to not include
        subset.pop()
        dfs(i + 1)
    
    dfs(0)
    return result


# Testing
nums = [1,2,3]
print(subsets(nums))

File: 10. Backtracking/078-subsets.go
// Leetcode 78

package main

func subsets(nums []int) [][]int {
	result := make([][]int, 0)
	subset := make([]int, 0)

	var dfs func(index int)
	dfs = func(index int) {
		result = append(result, append([]int{}, subset...))
		if index == len(nums) {
			return
		}
		for i := index; i < len(nums); i++ {
			subset = append(subset, nums[i])
			dfs(i + 1)
			subset = subset[:len(subset)-1]
		}
	}
	dfs(0)
	return result
}
